test/models/user_test.rb

```diff
@@ -990,4 +990,42 @@ def editing_self_helper
       end
     end
   end
+
+  describe '#visible_environments' do
+    teardown do
+      SETTINGS[:locations_enabled] = true
+      SETTINGS[:organizations_enabled] = true
+    end
+
+    test 'should check for environments permission with orgs/locs disabled' do
+      SETTINGS[:locations_enabled] = false
+      SETTINGS[:organizations_enabled] = false
+      setup_user 'view', 'environments', 'name != production'
+      all_but_production = Environment.unscoped.select { |e| e.name != 'production' }
+      assert_equal all_but_production.map(&:name), User.current.visible_environments
+    end
+
+    test 'should show the list of environments visible when only orgs are enabled' do
+      SETTINGS[:locations_enabled] = false
+      SETTINGS[:organizations_enabled] = true
+      Taxonomy.expects(:locations_enabled).returns(false).at_least_once
+      setup_user 'view', 'environments', 'name = test_env_org1'
+      FactoryGirl.create(:environment, :name => 'test_env_org1',
+                         :organizations => [User.current.organizations.first])
+      assert_equal ['test_env_org1'], User.current.visible_environments
+    end
+
+    test 'should show the list of environments visible when only locs are enabled' do
+      SETTINGS[:locations_enabled] = true
+      SETTINGS[:organizations_enabled] = false
+      setup_user 'view', 'environments', 'name = test_env_loc1'
+      FactoryGirl.create(:environment, :name => 'test_env_loc1',
+                         :locations => [User.current.locations.first])
+      assert_equal ['test_env_loc1'], User.current.visible_environments
+    end
+
+    test 'should show the list of environments visible when both orgs/locs are enabled' do
+      assert_equal ['production'], User.current.visible_environments
```

Comment:
There's a bug in the method (already existing prior to this patch, but I'll fix it). Locations and Organizations that have NO orgs or locs will never show up as visible environments, not even for the admin user. The reason is that there's a join/where on taxable_taxonomies that will not allow these to show up